File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        
        int n = s.length();
        stack<char> st;
        
        st.push(s[0]);
        
        for(int i = 1; i < n; i++)
        {
            if(s[i] == '(' || s[i] == '{' || s[i] == '[')
                st.push(s[i]);
            else if(s[i] == '}')
            {
                if(st.size() && st.top() == '{')
                    st.pop();
                else
                    return false;
            }
            else if(s[i] == ']')
            {
                if(st.size() && st.top() == '[')
                    st.pop();
                else
                    return false;
            }
            else if(s[i] == ')')
            {
                if(st.size() && st.top() == '(')
                    st.pop();
                else
                    return false;
            }
        }
        
        if(st.size())
            return false;
        return true;
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String str) {
        
        Stack<Character> st = new Stack<>();
        int n = str.length();
        
        for(int i = 0; i < n; i++)
        {
            char ch = str.charAt(i);
            
            // open brackets
            if(ch == '(') {
                st.push(ch);    
            }
            else if(ch == '[') {
                st.push(ch);  
            }
            else if(ch == '{') {
                st.push(ch);  
            }
            
            else if(ch == ')') {
                
                if(st.size() == 0)
                    return false;
                else if(st.peek() == '(')
                    st.pop();
                else {
                    return false;
                }
            }
            
            else if(ch == ']') {
                
                if(st.size() == 0)
                    return false;
                else if(st.peek() == '[')
                    st.pop();
                else {
                    return false;
                }
            }
            
            else if(ch == '}') {
                if(st.size() == 0)
                    return false;
                else if(st.peek() == '{')
                    st.pop();
                else {
                    return false;
                }
            }
        }
        
        return (st.size() == 0);
    }
}